solang-server/src/serverutils/backend.rs

```diff
@@ -0,0 +1,279 @@
+use jsonrpc_core::Result;
+use serde_json::Value;
+use tower_lsp::lsp_types::*;
+use tower_lsp::{Client, LanguageServer};
+
+use solang::file_cache::FileCache;
+use solang::parse_and_resolve;
+use solang::Target;
+
+use lsp_types::{Diagnostic, DiagnosticSeverity, Position, Range};
+use solang::sema::*;
+
+use std::path::PathBuf;
+
+#[derive(Debug, Default)]
+pub struct Backend {
+    state: Vec<usize>,
+}
+
+impl Backend {
+    // Calculate the line and coloumn from the Loc offset recieved from the parser
+    // Do a linear search till the correct offset location is matched
+    fn file_offset_to_line_column(data: &str, loc: usize) -> (usize, usize) {
+        let mut line_no = 0;
+        let mut past_ch = 0;
+
+        for (ind, c) in data.char_indices() {
+            if c == '\n' {
+                if ind == loc {
+                    break;
+                } else {
+                    past_ch = ind + 1;
+                    line_no += 1;
+                }
+            }
+            if ind == loc {
+                break;
+            }
+        }
+
+        (line_no, loc - past_ch)
+    }
+
+    // Convert the diagnostic messages recieved from the solang to lsp diagnostics types.
+    // Returns a vector of diagnostic messages for the client.
+    fn convert_to_diagnostics(ns: ast::Namespace, filecache: &mut FileCache) -> Vec<Diagnostic> {
+        let mut diagnostics_vec: Vec<Diagnostic> = Vec::new();
+
+        for diag in ns.diagnostics {
+            let pos = diag.pos.unwrap();
+
+            let fl = &ns.files[pos.0];
+
+            let file_cont = filecache.get_file_contents(fl.as_str());
+
+            let l1 = Backend::file_offset_to_line_column(&file_cont.as_str(), pos.1);
+
+            let l2 = Backend::file_offset_to_line_column(&file_cont.as_str(), pos.2);
+
+            let p1 = Position::new(l1.0 as u64, l1.1 as u64);
+
+            let p2 = Position::new(l2.0 as u64, l2.1 as u64);
+
+            let range = Range::new(p1, p2);
+
+            let diagnostic = &diag;
+
+            let sev = match diagnostic.level {
+                ast::Level::Info => DiagnosticSeverity::Information,
+                ast::Level::Warning => DiagnosticSeverity::Warning,
+                ast::Level::Error => DiagnosticSeverity::Error,
+                ast::Level::Debug => DiagnosticSeverity::Hint,
```

Comment:
debug is not a hint, debug diagnostics should be ignored.

Reply:
done